### PR TITLE
[Feature] Handle json payloads with schema registry

### PR DESCRIPTION
**Pull request by mproch**
Main change is new (de)serialization factory: `ConfluentJsonPayloadDeserializerFactory` and `ConfluentJsonPayloadSerializerFactory`, based on https://github.com/allegro/json-avro-converter.
In this case, Avro schema is still used, the change only causes different payload seralization

Also, refactor was made in Avro test setup to make handling new cases easier:
- split `KafkaAvroSpecMixin` to
  - `KafkaAvroSpecMixin`
  -  `SchemaRegistryMixin`
  -  `SchemaRegistryOperation`
- rewrite `ConfluentKafkaAvroDeserializationSpec` and `ConfluentKafkaAvroSerializationSpec` so that they can handle different payload serializations

### CODE REVIEW

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala [hunk 1]
```diff
@@ -13,21 +13,7 @@ trait SchemaRegistryClient extends Serializable {
...
-  /**
-    * Latest schema by subject - we assume there can be some latency
-    *
-    * @param topic
-    * @param isKey
-    * @return
-    */
-  def getLatestSchema(topic: String, isKey: Boolean): Validated[SchemaRegistryError, SchemaWithMetadata]
-
-  def getSchema(topic: String, version: Option[Int], isKey: Boolean): Validated[SchemaRegistryError, SchemaWithMetadata] =
```
Comment (lciolecki) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala:27]: Right now, we don't need it?

**Review by lciolecki — APPROVED** [review]
LGTM, but.. I see one big problem. Right now we can't use avro and json together. I mean we can't have different source/sink for json and avro, because SR always will return list for all subjects..

So, we should figure how to distinguish avro subject from json subject..

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/jsonpayload/ConfluentJsonPayloadDeserializerFactory.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    val specificClass = {
+      val clazz = implicitly[ClassTag[T]].runtimeClass
+      //This is a bit tricky, Allegro decoder requires SpecificRecordBase instead of SpecificRecord
+      if (classOf[SpecificRecordBase].isAssignableFrom(clazz)) Some(clazz.asInstanceOf[Class[SpecificRecordBase]]) else None
+    }
+
+    new ConfluentKafkaAvroDeserializer[T](kafkaConfig, schemaDataOpt.orNull,
+      schemaRegistryClient, isKey = false, specificClass.isDefined) {
+
+      private val converter = new JsonAvroConverter();
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/jsonpayload/ConfluentJsonPayloadDeserializerFactory.scala:33]: ConfluentKafkaAvroDeserializer is also not serializable - that's why we have Factory :)

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/helpers/SchemaRegistryOperations.scala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+import pl.touk.nussknacker.engine.avro.schemaregistry.confluent.ConfluentUtils
+import pl.touk.nussknacker.engine.avro.schemaregistry.confluent.serialization.{AbstractConfluentKafkaAvroDeserializer, AbstractConfluentKafkaAvroSerializer}
+import pl.touk.nussknacker.engine.flink.util.keyed.{KeyedValue, StringKeyedValue}
+import pl.touk.nussknacker.engine.kafka.{KafkaClient, KafkaZookeeperUtils}
+import pl.touk.nussknacker.engine.util.json.BestEffortJsonEncoder
+import pl.touk.nussknacker.test.PatientScalaFutures
+
+import java.nio.charset.StandardCharsets
+
+trait SchemaRegistryOperations extends Matchers with PatientScalaFutures {
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/helpers/SchemaRegistryOperations.scala:19]: yes, but they need SchemaRegistry - I'll change to KafkaWithSchemaRegistryOperations

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala (hunk 1 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala:27]: well - it comiles so I guess we don't need it :)

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/ConfluentSchemaRegistryProvider.scala [hunk 4]
```diff
@@ -39,21 +40,45 @@ object ConfluentSchemaRegistryProvider extends Serializable {
...
+  }
+
+  def jsonPayload(schemaRegistryClientFactory: ConfluentSchemaRegistryClientFactory,
+              processObjectDependencies: ProcessObjectDependencies,
+              formatKey: Boolean): ConfluentSchemaRegistryProvider = ConfluentSchemaRegistryProvider(schemaRegistryClientFactory,
+    new ConfluentJsonPayloadSerializerFactory(schemaRegistryClientFactory),
+    new ConfluentJsonPayloadDeserializerFactory(schemaRegistryClientFactory),
+    processObjectDependencies,
+    formatKey)
+
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/ConfluentSchemaRegistryProvider.scala:70]: ok...

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/helpers/SchemaRegistryOperations.scala [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    def apply(testName: String, schemas: List[Schema]): TopicConfig = {
+      val inputTopic = s"$inputPrefix.$kafkaTopicNamespace.$testName"
+      val outputTopic = s"$outputPrefix.$kafkaTopicNamespace.$testName"
+      new TopicConfig(inputTopic, outputTopic, schemas, isKey = false)
+    }
+  }
+
+}
+
+
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/helpers/SchemaRegistryOperations.scala:117]: ook

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/ConfluentKafkaAvroSeDeSpecMixin.scala [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        ConfluentSchemaRegistryProvider(MockSchemaRegistry.factory, testProcessObjectDependencies),
+        new SimpleKafkaAvroSerializer(MockSchemaRegistry.schemaRegistryMockClient),
+        new SimpleKafkaAvroDeserializer(MockSchemaRegistry.schemaRegistryMockClient, false))
+
+  lazy val jsonSetup: SchemaRegistryProviderSetup = SchemaRegistryProviderSetup("json",
+        ConfluentSchemaRegistryProvider.jsonPayload(MockSchemaRegistry.factory, testProcessObjectDependencies, formatKey = false),
+        SimpleKafkaJsonSerializer,
+        SimpleKafkaJsonDeserializer)
+
+  case class SchemaRegistryProviderSetup(name: String,
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/ConfluentKafkaAvroSeDeSpecMixin.scala]: ok :)

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/ConfluentUtils.scala [hunk 7]
```diff
@@ -47,4 +56,24 @@ object ConfluentUtils {
       .parsePayloadToByteBuffer(bytes)
       .valueOr(exc => throw new SerializationException(exc.getMessage, exc))
       .getInt
+
+  def typeInfoForSchema[T: ClassTag](kafkaConfig: KafkaConfig, schemaDataOpt: Option[RuntimeSchemaData]): TypeInformation[T] = {
+    val clazz = classTag[T].runtimeClass.asInstanceOf[Class[T]]
+    val isSpecificRecord = AvroUtils.isSpecificRecord[T]
+
+    schemaDataOpt match {
+      case Some(schema) if !isSpecificRecord && KryoGenericRecordSchemaIdSerializationSupport.schemaIdSerializationEnabled(kafkaConfig) =>
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/ConfluentUtils.scala]: Old code.. ;) Now I will do it in another way :p

**Review by arkadius — APPROVED** [review]

**Pull request merged**